x-pack/plugins/upgrade_assistant/common/types.ts

```diff
@@ -205,6 +203,21 @@ export interface EnrichedDeprecationInfo extends DeprecationInfo {
   blockerForReindexing?: 'index-closed'; // 'index-closed' can be handled automatically, but requires more resources, user should be warned
 }
 
+export interface MlAction {
```

Comment:
Do we expect to regularly maintain/add/remove some of these types more than others? `MLAction` for example -- will we eventually remove it, and do we expect to add other actions similar to it for other plugins?

In contrast, will we want to keep `ReindexAction`, `IndexSettingAction`, and others around for some time since they seem to form the core of Upgrade Assistant functionality?

If this is the case, would it make sense to start using folders to organize our types along these lines, as a way of making it easier to know which is which?

Reply:
That's a great point. I imagine some actions will end up being specific to a release. I agree it makes sense to start thinking how to better organize this.

@benwtrent can you comment specifically around the ML deprecation? Do you foresee this only to be applicable to 8.0?

Reply:
@alisonelizabeth 

🔮 🤔 🤯 

Probably :). But, I don't have a high level view of all of our plans. @droberts195 ? What say you?

Reply:
At every major version upgrade we would like to force users to either upgrade or delete really old ML job model snapshots.  So it's very similar to forcing people to move to the newer Lucene index formats via reindexing.

There might be one-off ML upgrade actions _as well_ for some particular release in the future, but I think it's safe to say ML model snapshot upgrade is going to stay forever now.

Reply:
> do we expect to add other actions similar to it for other plugins?

Maybe, but it's important to realise that ML is an order of magnitude bigger and more complex than the average plugin.

Possibly one day there might be something around moving rollups v1 configs to rollups v2 configs?  Or maybe changing something in transforms configs in some future release.  Or changing some aspect of the security config.  They sound like one-off things that would only be required for one major version upgrade.  So it may well make sense to separate out one-off migrations from perennial migrations.

But ML model state can be huge and complex and will be with us forever (unless somebody decides to scrap the anomaly detection feature, which seems unlikely).  As the security and observability solutions utilise anomaly detection more and more it's becoming a key building block for the whole Elastic product lineup, and makes sense to treat almost on a par with indices in terms of the importance of not accidentally losing it on upgrade.

Reply:
> perennial migrations

I'm pretty sure I planted these in my garden last month! 🌷 

Jokes aside, thanks for the clarification @droberts195. Looking at ML in the light you've cast, I can see how it makes sense to treat it as a permanent UA concern. 👍

Reply:
Thanks for the details @droberts195. Very helpful!

@cjcenizal I think I’ll keep the current structure in place for now. However, I definitely see how it might be beneficial to reorganize if we do add additional actions that are more focused to a specific release. I’ll keep this in mind in the future.